src/txHashBuilder.c

```diff
@@ -1466,9 +1545,168 @@ static void txHashBuilder_assertCanLeaveNetworkId(tx_hash_builder_t* builder)
+
+	txHashBuilder_assertCanLeaveCollateralReturn(builder);
+	ASSERT(builder->includeTotalCollateral);
+
+
+	// add TotalCollateral item into the main tx body map
+	BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, TX_BODY_KEY_TOTAL_COLLATERAL);
+	BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, txColl);
+
+	builder->state = TX_HASH_BUILDER_IN_TOTAL_COLLATERAL;
+}
+
+static void txHashBuilder_assertCanLeaveTotalCollateral(tx_hash_builder_t* builder)
+{
+	_TRACE("state = %d", builder->state);
+
+	switch (builder->state) {
+	case TX_HASH_BUILDER_IN_TOTAL_COLLATERAL:
+		break;
+	case TX_HASH_BUILDER_IN_COLLATERAL_RETURN:
+	case TX_HASH_BUILDER_IN_NETWORK_ID:
+	case TX_HASH_BUILDER_IN_REQUIRED_SIGNERS:
+	case TX_HASH_BUILDER_IN_COLLATERALS:
+	case TX_HASH_BUILDER_IN_SCRIPT_DATA_HASH:
+	case TX_HASH_BUILDER_IN_MINT:
+	case TX_HASH_BUILDER_IN_VALIDITY_INTERVAL_START:
+	case TX_HASH_BUILDER_IN_AUX_DATA:
+	case TX_HASH_BUILDER_IN_WITHDRAWALS:
+	case TX_HASH_BUILDER_IN_CERTIFICATES:
+	case TX_HASH_BUILDER_IN_TTL:
+	case TX_HASH_BUILDER_IN_FEE:
+		txHashBuilder_assertCanLeaveRequiredSigners(builder);
+		ASSERT(!builder->includeTotalCollateral);
+		break;
+
+	default:
+		ASSERT(false);
+	}
+}
+
+void txHashBuilder_enterReferenceInputs(tx_hash_builder_t* builder )
+{
+	_TRACE("state = %d", builder->state);
+
+	txHashBuilder_assertCanLeaveTotalCollateral(builder);
+	// we don't allow an empty list for an optional item
+	ASSERT(builder->remainingReferenceInputs > 0);
+	{
+		// Enter reference inputs
+		BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, TX_BODY_KEY_REFERENCE_INPUTS);
+		BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, builder->remainingReferenceInputs);
+	}
+	builder->state = TX_HASH_BUILDER_IN_REFERENCE_INPUTS;
+
+}
+void txHashBuilder_addReferenceInputs(
+        tx_hash_builder_t* builder,
+        const uint8_t* utxoHashBuffer, size_t utxoHashSize,
+        uint32_t utxoIndex
+)
+{
+	_TRACE("state = %d, remainingInputs = %u", builder->state, builder->remainingReferenceInputs);
+
+	ASSERT(utxoHashSize < BUFFER_SIZE_PARANOIA);
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_REFERENCE_INPUTS);
+	ASSERT(builder->remainingReferenceInputs > 0);
+	builder->remainingReferenceInputs--;
+
+	cbor_append_txInput(builder, utxoHashBuffer, utxoHashSize, utxoIndex);
+}
+
+
+static void txHashBuilder_assertCanLeaveReferenceInputs(tx_hash_builder_t* builder)
+{
+	_TRACE("state = %d", builder->state);
+
+	switch (builder->state) {
+	case TX_HASH_BUILDER_IN_REFERENCE_INPUTS:
+		break;
+	case TX_HASH_BUILDER_IN_TOTAL_COLLATERAL:
+	case TX_HASH_BUILDER_IN_COLLATERAL_RETURN:
+	case TX_HASH_BUILDER_IN_NETWORK_ID:
+	case TX_HASH_BUILDER_IN_REQUIRED_SIGNERS:
+	case TX_HASH_BUILDER_IN_COLLATERALS:
+	case TX_HASH_BUILDER_IN_SCRIPT_DATA_HASH:
+	case TX_HASH_BUILDER_IN_MINT:
+	case TX_HASH_BUILDER_IN_VALIDITY_INTERVAL_START:
+	case TX_HASH_BUILDER_IN_AUX_DATA:
+	case TX_HASH_BUILDER_IN_WITHDRAWALS:
+	case TX_HASH_BUILDER_IN_CERTIFICATES:
+	case TX_HASH_BUILDER_IN_TTL:
+	case TX_HASH_BUILDER_IN_FEE:
```

Comment:
This is becoming annoying. Give it a while and try to come up with some reasonable solution to avoid these long lists in `txHashBuilder_assertCanLeave*` functions (we should discuss it before coding it). Perhaps using comparisons between states?... Inequalities are implicit and so more error-prone vs. explicit lists (e.g. when we add states), but these lists are becoming too long to read/check, and we don't expect to put states in between the existing ones.